src/collections/arc_slice.c

```diff
@@ -0,0 +1,84 @@
+//
+// Copyright (c) 2024 ZettaScale Technology
+//
+// This program and the accompanying materials are made available under the
+// terms of the Eclipse Public License 2.0 which is available at
+// http://www.eclipse.org/legal/epl-2.0, or the Apache License, Version 2.0
+// which is available at https://www.apache.org/licenses/LICENSE-2.0.
+//
+// SPDX-License-Identifier: EPL-2.0 OR Apache-2.0
+//
+// Contributors:
+//   ZettaScale Zenoh Team, <zenoh@zettascale.tech>
+//
+
+#include "zenoh-pico/collections/arc_slice.h"
+
+#include <assert.h>
+#include <stddef.h>
+#include <stdio.h>
+#include <string.h>
+
+_z_arc_slice_t _z_arc_slice_empty(void) {
+    _z_arc_slice_t s;
+    s.len = 0;
+    s.start = 0;
+    s.slice.in = NULL;
+    return s;
+}
+
+_z_arc_slice_t _z_arc_slice_wrap(_z_slice_t s, size_t offset, size_t len) {
+    assert(offset + len <= s.len);
+
+    _z_arc_slice_t arc_s;
+    arc_s.slice = _z_slice_rc_new_from_val(s);
+    arc_s.len = len;
+    arc_s.start = offset;
+    return arc_s;
+}
+
+_z_arc_slice_t _z_arc_slice_get_subslice(const _z_arc_slice_t* s, size_t offset, size_t len) {
+    assert(offset + len <= s->len);
+    assert(s->slice.in != NULL || (len == 0 && offset == 0));
```

Comment:
Asserts will be removed on `RELEASE` compile, I'd say we should use them only in tests.

Reply:
This is internal function, thus I believe asserts here are justified, since they clearly represent the contract on input arguments (i.e. under no circumstances they are supposed to be violated, no matter what the user input would be to the public api).

Reply:
But if the check is important, `assert` is not ideal:
* They disappear when `NDEBUG` is defined.
* They don't make much sense on bare metal systems and as such might be unavailable.

So if the error is somewhat recoverable, then the function should return an error which then propagate or is processed.

If it's truly unrecoverable, I guess we need to create a mechanism for that.

Reply:
Yes, this error is supposed to be non-recoverable (but likely will be caught before in some higher level api and bubbled up through error propagation mechanism). But as I said it is internal function and thus asserts are supposed to help catching this violation during testing/development. Using asserts is really a standard practice in specifying contracts, I do not see why we should complicate it.

Reply:
Ok I didn't understand the intent correctly. if it's used as in-place testing, then it's fine. 

Maybe it could be worth to wrap them, in case we stumble upon an availability issue on some platforms but it's not in the scope of this PR.